benchmark/framework/cli.py

```diff
@@ -0,0 +1,150 @@
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Cli implementation for performance tests suits."""
+import ast
+import inspect
+from typing import Any, Callable, Dict, List, Optional
+
+import click
+from click.core import Argument, Command, Context, Option, Parameter
+
+from .executor import Executor
+from .func_details import FuncDetails
+from .report_printer import PerformanceReportPrinter
+
+
+class DefaultArgumentsMultiple(Argument):
+    """Multiple arguments with default value."""
+
+    def __init__(self, *args, **kwargs):
+        """Create MultipleArguments instance."""
+        kwargs["nargs"] = -1
+        default = kwargs.pop("default", tuple())
+        super().__init__(*args, **kwargs)
+        self.default = default
+
+    def full_process_value(self, ctx: Context, value: Any) -> Any:
+        """Given a value and context this runs the logic to convert the value as necessary."""
+        return super().process_value(ctx, value or self.default)
+
+
+class TestCli:
+    """Performance test client."""
+
+    def __init__(
+        self,
+        func: Callable,
+        executor_class=Executor,
+        report_printer_class=PerformanceReportPrinter,
+    ):
+        """
+        Make performance client.
+
+        func - should be function with first parameter multithreading.Queue.
+        func  - should have docstring, and default values for every extra argument.
+        """
+        self.func_details = FuncDetails(func)
+        self.func_details.check()
+        self.func = func
+        self.executor_class = Executor
+        self.report_printer_class = report_printer_class
+
+    def _make_command(self) -> Command:
+        """Make  cli.core.Command."""
+        return Command(
+            None,  # type: ignore
+            params=self._make_command_params(),
+            callback=self._callback,
+            help=self._make_help(),
+        )
+
+    def _make_command_params(self) -> Optional[List[Parameter]]:
+        """Make parameters and arguments for cli.Command."""
+        return list(self.executor_params().values()) + [self._function_args()]
+
+    def _make_help(self) -> str:
+        """Make help for command."""
+        return inspect.cleandoc(
+            f"""
+        {self.func_details.doc}
+
+        ARGS is function arguments in format: `{','.join(self.func_details.argument_names)}`
+
+        default ARGS is `{self.func_details.default_argument_values_as_string}`
+        """
+        )
+
+    def executor_params(self) -> Dict[str, Parameter]:
+        """Get parameters used by Executor."""
+        return {
```

Comment:
We prefer assignment to a helper variable and then return statement on new line, so:`
```
some_var = { .... lots of multiline content ... }
return some_var
```
It makes for slightly more explicit code.

Reply:
done